### PR TITLE
Implement indexed mode

### PR DESCRIPTION
**Pull request by tmcw**
Test for #177

I have some hacky benchmarks to push as well, which I should. It looks like modeFast is reliably faster than mode, but around 30%

### CODE REVIEW

File: src/mode_fast.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    var mode/*: number */;
+    var modeCount;
+
+    index.forEach(function (count, key) {
+        if (modeCount === undefined || modeCount < count) {
+            modeCount = count;
+            mode = key;
+        }
```
Comment (mourner) [inline src/mode_fast.js]: I think it would be simpler to move this logic to the loop above.

Comment (tmcw) [inline src/mode_fast.js]: Good idea, the benchmarks tell me it's faster too. Implemented in last commit.

File: src/mode_fast.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    // them like
+    // { value: count }
+    var index = new Map();
+
+    // A running `mode` and the number of times it has been encountered.
+    var mode;
+    var modeCount = 0;
+
+    for (var i = 0; i < x.length; i++) {
+        var newCount = index.get(x[i]);
```
Comment (tmcw) [inline src/mode_fast.js]: Perf note: was wondering if the lookup of `x[i]` these three times could be made faster by assigning `x[i]` to a variable. Tried it and benchmarks didn't show any evidence of a speedup.

File: src/mode_fast.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    // This index will reflect the incidence of different values, indexing
+    // them like
+    // { value: count }
+    var index = new Map();
+
+    // A running `mode` and the number of times it has been encountered.
+    var mode;
+    var modeCount = 0;
+
+    for (var i = 0; i < x.length; i++) {
```
Comment (tmcw) [inline src/mode_fast.js]: Perf note: was wondering if caching `x.length` would give a detectable speedup. Tried it, and it didn't.

**Pull request merged**